runtime/cudaq/builder/kernels.cpp

```diff
@@ -0,0 +1,188 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "kernels.h"
+#include "common/EigenDense.h"
+#include <iostream>
+
+namespace cudaq::details {
+
+std::vector<std::string> grayCode(std::size_t rank) {
+  std::function<void(std::vector<std::string> &, std::size_t)> grayCodeRecurse;
+  grayCodeRecurse = [&grayCodeRecurse](std::vector<std::string> &g,
+                                       std::size_t rank) {
+    auto k = g.size();
+    if (rank <= 0)
+      return;
+
+    for (int i = k - 1; i >= 0; --i) {
+      auto c = "1" + g[i];
+      g.push_back(c);
+    }
+    for (int i = k - 1; i >= 0; --i) {
+      g[i] = "0" + g[i];
+    }
+
+    grayCodeRecurse(g, rank - 1);
+  };
+
+  std::vector<std::string> g{"0", "1"};
+  grayCodeRecurse(g, rank - 1);
+
+  return g;
+}
+
+std::vector<std::size_t> getControlIndices(std::size_t grayRank) {
+  auto code = grayCode(grayRank);
+  std::vector<std::size_t> ctrlIds;
+  for (std::size_t i = 0; i < code.size(); i++) {
+    auto a = std::stoi(code[i], nullptr, 2);
+    auto b = std::stoi(code[(i + 1) % code.size()], nullptr, 2);
+    auto c = a ^ b % code.size();
+    ctrlIds.emplace_back(std::log2(c));
+  }
+  return ctrlIds;
+}
+
+std::size_t mEntry(std::size_t row, std::size_t col) {
+  auto b_and_g = row & ((col >> 1) ^ col);
+  std::size_t sum_of_ones = 0;
+  while (b_and_g > 0) {
+    if (b_and_g & 0b1)
+      sum_of_ones += 1;
+
+    b_and_g = b_and_g >> 1;
+  }
+  return std::pow(-1, sum_of_ones);
+}
+
+std::vector<double> computeAngle(std::vector<double> alpha) {
```

Comment:
Does `alpha` scale ~ O(2^N)? Maybe we should use `const std::vector<double>&` to prevent copies.